universe/NamedValueRefManager.cpp

```diff
@@ -0,0 +1,190 @@
+#include "ValueRefs.h"
+
+#include <functional>
+#include <iomanip>
+#include <iterator>
+#include <mutex>
+#include <unordered_map>
+#include <boost/algorithm/string/classification.hpp>
+#include <boost/algorithm/string/predicate.hpp>
+#include <boost/algorithm/string/split.hpp>
+#include <boost/log/expressions.hpp>
+#include <boost/range/adaptor/filtered.hpp>
+#include <boost/range/adaptor/map.hpp>
+#include <boost/range/numeric.hpp>
+#include "Enums.h"
+#include "Field.h"
+#include "Fighter.h"
+#include "Fleet.h"
+#include "Pathfinder.h"
+#include "Planet.h"
+#include "ShipDesign.h"
+#include "ShipHull.h"
+#include "ShipPart.h"
+#include "Ship.h"
+#include "Species.h"
+#include "System.h"
+#include "Tech.h"
+#include "UniverseObjectVisitors.h"
+#include "UniverseObject.h"
+#include "Universe.h"
+#include "../Empire/EmpireManager.h"
+#include "../Empire/Empire.h"
+#include "../Empire/Supply.h"
+#include "../util/GameRules.h"
+#include "../util/Logger.h"
+#include "../util/MultiplayerCommon.h"
+#include "../util/Random.h"
+
+namespace expr = boost::log::expressions;
+
+NamedValueRefManager* NamedValueRefManager::s_instance = nullptr;
+
+NamedValueRefManager::NamedValueRefManager() {
+    if (s_instance)
+        throw std::runtime_error("Attempted to create more than one NamedValueRefManager.");
+
+    // Only update the global pointer on sucessful construction.
+    InfoLogger() << "NamedValueRefManager::NameValueRefManager constructs singleton " << this;
+    s_instance = this;
+}
+
+template <typename V>
+V* const NamedValueRefManager::GetValueRefImpl(std::map<NamedValueRefManager::key_type, std::unique_ptr<V>>& registry, const std::string& label, const std::string& name) /*const*/ {
```

Comment:
probably don't need the `label` param since the same info is provided by the type `V`, but not high priority to rework...

Reply:
also it is hard to express that the generic registry gets used instead of the int and double ones (the entries in the generic registry are all ValueRefBase which are not ValueRef<int> or ValueRef<double>).